internal/http/interceptors/auth/auth_test.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2018-2020 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package auth
+
+import (
+	"testing"
+)
+
+func TestApplyWWWAuthenticate(t *testing.T) {
+	type test struct {
+		userAgent            string
+		userAgentMap         map[string]string
+		availableCredentials []string
+		expected             []string
+	}
+
+	tests := []*test{
+		// no user agent we return all available credentials
+		&test{
+			userAgent:            "",
+			userAgentMap:         map[string]string{},
+			availableCredentials: []string{},
+			expected:             []string{},
+		},
+
+		// no user map we return all available credentials
+		&test{
+			userAgent:            "mirall",
+			userAgentMap:         map[string]string{},
+			availableCredentials: []string{"basic"},
+			expected:             []string{"basic"},
+		},
+
+		// user agent set but no mapping set we return all credentials
+		&test{
+			userAgent:            "mirall",
+			userAgentMap:         map[string]string{},
+			availableCredentials: []string{"basic"},
+			expected:             []string{"basic"},
+		},
+
+		// user mapping set to non available credential, we return all available
+		&test{
+			userAgent:            "mirall",
+			userAgentMap:         map[string]string{"mirall": "notfound"},
+			availableCredentials: []string{"basic", "bearer"},
+			expected:             []string{"basic", "bearer"},
+		},
+
+		// user mapping set and we return only desired credential
+		&test{
+			userAgent:            "mirall",
+			userAgentMap:         map[string]string{"mirall": "bearer"},
```

Comment:
this does not allow for multiple credentials, right? The case:

```golang
map[string]string{
	"mirall": "basic",
	"mirall": "bearer",
}
```

is it meant to work like this? Since Go doesn't allow for duplicated keys on a map; or do we want to limit a user agent multiplex to one strategy only?

Reply:
Alternatively, we could use POXIX compliant regex for the user agent

Reply:
Yes, one user agent with one challenge. If need arises we can change it later for a map[string][]string{} to support multiple auth methods per user-agent.